# MNIST using Neural Networks

Run the statement to import the TensorFlow library:

In [123]:
import tensorflow as tf

Run the statement to load and prepare the [MNIST dataset](http://yann.lecun.com/exdb/mnist/) and to convert the samples from integers to floating-point numbers:

In [124]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
x_train = x_train.reshape(-1, 28, 28, 1)
x_test = x_test.reshape(-1, 28, 28, 1)

Build the model with a single layer and relu activation function:

In [177]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(14, (3, 3), activation='relu', input_shape=(28, 28, 1)),  
  tf.keras.layers.MaxPooling2D((2, 2)),                                 

  tf.keras.layers.Flatten(),

  tf.keras.layers.Dense(20, activation='relu'),
  tf.keras.layers.Dense(10),
  
])
# try to add convolutions and change the no of layers and no of nodes/neurons
# to add a new layer
# input_shape=(28, 28, 1)
# tf.keras.layers.Dense(10, activation='relu') 
# to add convolutions 
# tf.keras.layers.Conv2D(28, (3, 3), activation='relu', input_shape=(28, 28)) 
# tf.keras.layers.MaxPooling2D((2, 2))

Choosing our loss function, in this case crossentropy:

In [178]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [179]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

Adjusting the model parameters to minimize the loss of our model: 

In [180]:
model.fit(x_train, y_train, epochs=2)

Epoch 1/2
1875/1875 [==============================] - 19s 10ms/step - loss: 0.2424 - accuracy: 0.9292
Epoch 2/2
1875/1875 [==============================] - 19s 10ms/step - loss: 0.0850 - accuracy: 0.9748


Checking the model's performance on the Validation-set.

In [181]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 1s - loss: 0.0662 - accuracy: 0.9782


[0.06617061793804169, 0.9782000184059143]

Try to improve the accuracy to 97% by changing the number of layers and/or adding convolutions.